### PR TITLE
Stringify keys of `other` hash when performing a merge

### PR DESCRIPTION
**Pull request by DRBragg**
### Summary

We’re given a couple of methods to modify the `ActionDispatch::Request::Session` object. All work as expected except `merge!`:

```ruby
session.to_hash
#=> {"session_id"=>"abc123"}
session.update({ foo: "bar" })
#=> {"session_id"=>"abc123", "foo" => "bar"}
session[:foo]
#=> "bar"
session[:bar] = "baz"
#=> {"session_id"=>"abc123", "foo" => "bar", "bar" => "baz"}
session[:bar]
#=> "baz"
session.merge!({ fizz: "buzz" })
#=> {"session_id"=>"abc123", "foo" => "bar", "bar" => "baz", fizz: "buzz"}
session[:fizz]
#=> nil
```

While `[]=` and `update` stringify the key before updating the “hash”, `merge!` does not. The session is properly loaded with stringified keys on the next request but within the request that the session is updated attempting to access by key will fail.

I ran into an issue when trying to write a failing test for this because the tests use `ActionController::TestSession` not `ActionDispatch::Request::Session`.  When you call `merge!` on this type of session it uses code in `rack/lib/rack/session/abstract/id.rb` which is already stringifying the keys. EDIT: And it turns out I'm an idiot and was running the wrong test 🤦‍♂️ 

Basically, this PR just mimics the same behavior as rack by stringifying the keys of the `other` hash passed into `merge!`

### CODE REVIEW

**Comment by DRBragg** [comment]
Sorry... Turns out I was running the wrong group of tests...  Update with correct code inbound.

**Comment by kamipo** [comment]
How about just `alias merge! update`? Since `Hash#update` and `Hash#merge!` is the same implementation in MRI.

https://github.com/ruby/ruby/blob/c26e41c9a324259a47c503addacd98f07cf2fb7b/hash.c#L7204
https://github.com/ruby/ruby/blob/c26e41c9a324259a47c503addacd98f07cf2fb7b/hash.c#L7206

**Comment by DRBragg** [comment]
> How about just `alias merge! update`? Since `Hash#update` and `Hash#merge!` is the same implementation in MRI.
> 
> [ruby/ruby@`c26e41c`/hash.c#L7204](https://github.com/ruby/ruby/blob/c26e41c9a324259a47c503addacd98f07cf2fb7b/hash.c#L7204) [ruby/ruby@`c26e41c`/hash.c#L7206](https://github.com/ruby/ruby/blob/c26e41c9a324259a47c503addacd98f07cf2fb7b/hash.c#L7206)

It looks like the primary use of `merge!` is to merge 2 sessions together. I actually messed this up in my first commit, by treating `other` like a `Hash` when it's actually another `ActionDispatch::Request::Session` instance. If we aliased `merge!` to `update` we'd run into the same issue I originally introduced where `ActionDispatch::Request::Session` doesn't respond to `stringify_keys`. This is why I updated `merge!` to first convert (or ensure) `other` is a hash before calling `stringify_keys`.

**Comment by byroot** [comment]
@DRBragg that doesn't change the fact that `Hash#update` and `Hash#merge!` are just aliases. So `Session#update` and `Session#merge!` should be aliases as well.

As for the casting with `.to_h`, that's another issue entirely and I'd suggest opening another PR for it, as some research should be done on what the exact casting method should be used (I'm not convinced `to_h` is the better one).

**Comment by byroot** [comment]
Actually `Kernel.Hash` is a bit too permissive, so we might need to do our own `if arg.respond_to?(:to_hash) ? arg.to_hash : raise TypeError ...`

**Comment by DRBragg** [comment]
@byroot Thanks for the feedback! I think I'm tracking. What do you think about something to the effect of:
```ruby
def update(hash)
  raise TypeError.new("Whatever we want to say") unless hash.respond_to?(:to_hash)

  load_for_write!
  @delegate.update hash.to_hash.stringify_keys
end
alias :merge! :update
```
If I'm understanding you correctly, that should take of what you and @kamipo are asking for. I think.

**Comment by byroot** [comment]
@DRBragg yes. However I'd rather split the two if you don't mind.

If you do a PR with just `alias :merge! :update` I can merge it right away, it's a nobrainer.

Whereas for the `to_hash` part, I'll need to dig more into it, so I probably won't get to it until Monday, and right now I can't say for certain wether the current behavior is expected or not, that require some research.

All that to say that it's best to do two PRs.

**Comment by DRBragg** [comment]
> @DRBragg yes. However I'd rather split the two if you don't mind.
> 
> If you do a PR with just `alias :merge! :update` I can merge it right away, it's a nobrainer.
> 
> Whereas for the `to_hash` part, I'll need to dig more into it, so I probably won't get to it until Monday, and right now I can't say for certain wether the current behavior is expected or not, that require some research.
> 
> All that to say that it's best to do two PRs.

I understand, however I hesitate to do a PR only aliasing `merge!` to `update` until we're sure how to handle a `ActionDispatch::Request::Session` being passed in. Simply aliasing will cause a fair amount tests to break and potentially  any apps testing against `main`. If you feel it's best to get it in now I'll of course update the PR now but I'm also not in a rush and can wait to discuss further on Monday.

**Comment by byroot** [comment]
Oh you mean `merge!(Session)` is currently tested? If so then yeah forget what I said.

Go with `raise TypeError unless arg.respond_to?(:to_hash)`. I'll review this a bit more on Monday.

**Comment by DRBragg** [comment]
Yes! I'm sorry if that was unclear... I'll update the PR and we chat more on Monday. Thanks!

File: actionpack/lib/action_dispatch/request/session.rb [hunk 1]
```diff
@@ -176,9 +176,12 @@ def to_hash
       #   session.to_hash
       #   # => {"session_id"=>"e29b9ea315edf98aad94cc78c34cc9b2", "foo" => "bar"}
       def update(hash)
+        raise TypeError.new("#{hash.inspect} cannot being converted into a Hash") unless hash.respond_to?(:to_hash)
```
Comment (byroot) [inline actionpack/lib/action_dispatch/request/session.rb:179]: Please use the same error message than Ruby:

```ruby
>> {}.merge(1)
(irb):1:in `merge': no implicit conversion of Integer into Hash (TypeError)
```

so:

```ruby
unless hash.respond_to?(:to_hash)
  raise TypeError, "no implicit conversion of #{hash.class.name} into Hash"
end
```

**Review by byroot — APPROVED** [review]
Please squash all commits into one and then it's good for me.

**Comment by byroot** [comment]
Merged as 6a55aff4a9fc5021066146f2c4abebb98a8d9894.

Thanks!
